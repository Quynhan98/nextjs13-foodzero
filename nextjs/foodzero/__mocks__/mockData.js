export const AVATAR_MOCK =
  'https://firebasestorage.googleapis.com/v0/b/image-blog-155e2.appspot.com/o/foodzero%2Fjohn-doe.webp?alt=media&token=6963f744-1ab8-4b09-971d-2486cd5375bf'

export const PRICE_LIST_MOCK = {
  name: 'Deep Sea Snow White Cod Fillet',
  price: 20,
  description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
}

export const CARD_BLOG = {
  imageUrl:
    'https://firebasestorage.googleapis.com/v0/b/image-blog-155e2.appspot.com/o/foodzero%2FFeature%20Image.webp?alt=media&token=0a725961-6c86-4dcd-ab48-e5d7e2eb48fd',
  title: 'Fruit and vegetables and protection against diseases',
  description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
  category: 'Fashion',
  srcAvatar: AVATAR_MOCK,
  name: 'Julie Christie',
  date: '2022-12-06T23:29:35.000Z',
  comments: 2,
}

export const CARD_FEATURE = {
  title: 'Premium Quality',
  imageUrl:
    'https://firebasestorage.googleapis.com/v0/b/image-blog-155e2.appspot.com/o/foodzero%2FCircleIcon.webp?alt=media&token=1df373f3-27a3-45f6-82fa-d997c47c718f',
  description:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Neque congue arcu',
}

export const CARD_CATEGORY = {
  href: '/',
  category: 'Starters',
  imageUrl:
    'https://firebasestorage.googleapis.com/v0/b/image-blog-155e2.appspot.com/o/foodzero%2Fcategory.webp?alt=media&token=4e7dcf6c-2a66-4695-917d-af1d578d9a30',
}

export const USERS_MOCK = [
  {
    id: 'gff1',
    email: 'foodzero@gmail.com',
    password: '123456789',
    reservations: [],
  },
]

export const QUOTE_MOCK = [
  {
    id: '1',
    authorName: 'John Doe',
    job: 'Blogger',
    quote:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Purus lorem id penatibus impedient. Turpis egestas ultricies purus  Lorem ipsum dolor sit amet.',
    avatar:
      'https://firebasestorage.googleapis.com/v0/b/image-blog-155e2.appspot.com/o/foodzero%2Fjohn-doe.webp?alt=media&token=6963f744-1ab8-4b09-971d-2486cd5375bf',
  },
  {
    id: '2',
    authorName: 'Julie Christie',
    job: 'Blogger',
    quote:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Purus lorem id penatibus impedient. Turpis egestas ultricies purus  Lorem ipsum dolor sit amet1.',
    avatar:
      'https://firebasestorage.googleapis.com/v0/b/image-blog-155e2.appspot.com/o/foodzero%2Funnamed.webp?alt=media&token=f2607e25-62ae-4ed9-9f53-bb76e6f830fb',
  },
  {
    id: '3',
    authorName: 'Dianne Russell',
    job: 'Blogger',
    quote:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Purus lorem id penatibus impedient. Turpis egestas ultricies purus  Lorem ipsum dolor sit amet2.',
    avatar:
      'https://firebasestorage.googleapis.com/v0/b/image-blog-155e2.appspot.com/o/foodzero%2Fcute-avatar.webp?alt=media&token=9d70884c-e075-4f36-bf67-b196493badb8',
  },
]

export const OUR_MENU_MOCK = [
  {
    id: '1',
    name: 'Deep Sea Snow White Cod Fillet',
    price: 20,
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit1.',
  },
  {
    id: '2',
    name: 'Steak With Rosemary Butter',
    price: 22,
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit2.',
  },
  {
    id: '3',
    name: 'Cucumber Salad',
    price: 18,
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit3.',
  },
  {
    id: '4',
    name: 'Natural Wine Pairing',
    price: 90,
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit4.',
  },
]

export const BLOG_SECTION_MOCK = [
  {
    imageUrl:
      'https://firebasestorage.googleapis.com/v0/b/image-blog-155e2.appspot.com/o/foodzero%2FFeature%20Image.webp?alt=media&token=766a34ea-32e0-40e7-90e6-299f9af22eb1',
    title: 'Fruit and vegetables and protection against diseases',
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit1.',
    category: 'Fashion',
    srcAvatar:
      'https://firebasestorage.googleapis.com/v0/b/image-blog-155e2.appspot.com/o/foodzero%2Fjulie.webp?alt=media&token=c429d5f2-6f41-4233-970b-b2739746f7f1',
    name: 'Julie Christies',
    date: '2021-10-17T23:29:35.000Z',
    comments: 2,
    id: '1',
  },
  {
    imageUrl:
      'https://firebasestorage.googleapis.com/v0/b/image-blog-155e2.appspot.com/o/foodzero%2Fasparagus.webp?alt=media&token=d56ab468-1479-4181-844d-f5b231742244',
    title: "Asparagus Spring Salad with Rocket, Goat's Cheese",
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit2.',
    category: 'Fashion',
    srcAvatar:
      'https://firebasestorage.googleapis.com/v0/b/image-blog-155e2.appspot.com/o/foodzero%2Fdianne.webp?alt=media&token=ce3d7213-72ce-4caa-b43c-f6f07bb3b5b7',
    name: 'Dianne Russell',
    date: '2021-10-17T23:29:35.000Z',
    comments: 2,
    id: '2',
  },
]

export const MENU_MOCK = {
  id: '1',
  ourMenu: OUR_MENU_MOCK,
  starters: OUR_MENU_MOCK,
  mains: OUR_MENU_MOCK,
  pastriesAndDrinks: OUR_MENU_MOCK,
}
