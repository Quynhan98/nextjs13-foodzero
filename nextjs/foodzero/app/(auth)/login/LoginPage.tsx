'use client'

import React, { ChangeEvent, FormEvent, useCallback, useState } from 'react'
import { useRouter } from 'next/navigation'
import { Box, useToast } from '@chakra-ui/react'

// Component
import LoginForm from '@components/LoginForm'
import LoadingIndicator from '@components/LoadingIndicator'

// Types
import { LoginAccount } from '@self-types/index'

// Utils
import { loginValidate } from '@utils/validation'

// Hooks
import { useAuthContext } from '@hooks/useAuthContext'
import { useLoadingContext } from '@hooks/useLoadingContext'

// Constants
import { PAGE_URL, SERVER_ERROR } from '@constants/index'

// Services
import { IAuthContext } from '@self-types/AuthContext'

enum ErrorField {
  Email = 'email',
  Password = 'password',
}

const loginInit: LoginAccount = {
  email: '',
  password: '',
}

const initErrorMsgs = {
  email: '',
  password: '',
}

const LoginPage = () => {
  const router = useRouter()
  const toast = useToast()
  const { login } = useAuthContext() as IAuthContext
  const { setLoading, loading } = useLoadingContext()
  const [loginAccount, setLoginAccount] = useState(loginInit)
  const [errorMsgs, setErrorMsgs] = useState(initErrorMsgs)

  // Check disable button login
  const isDisableButton = !loginAccount.email || !loginAccount.password

  // Get input value
  const handleChange = useCallback((e: ChangeEvent<HTMLInputElement>): void => {
    const inputValues = { [e.target.name]: e.target.value }

    setLoginAccount((prev) => ({
      ...prev,
      ...inputValues,
    }))

    setErrorMsgs(initErrorMsgs)
  }, [])

  // Function submit login form
  const handleSubmitForm = useCallback(
    async (e: FormEvent<HTMLFormElement>) => {
      e?.preventDefault()

      const formatValidation = loginValidate(loginAccount)

      if (!formatValidation.isValid && formatValidation.error) {
        setErrorMsgs(formatValidation.error)
      } else {
        try {
          setLoading(true)
          const response = await login(loginAccount)

          if (response.id) {
            router.push(PAGE_URL.HOME.URL)
          }

          const responseError = response as unknown as {
            field: ErrorField
            message: string
          }

          if (responseError.field) {
            const errorField = {
              email:
                responseError.field === ErrorField.Email
                  ? responseError.message
                  : '',
              password:
                responseError.field === ErrorField.Password
                  ? responseError.message
                  : '',
            }

            setErrorMsgs((prev) => ({
              ...prev,
              ...errorField,
            }))
          }
        } catch (error) {
          toast({
            title: 'Error',
            description: SERVER_ERROR,
            status: 'error',
            isClosable: true,
            position: 'top-right',
          })
        } finally {
          setLoading(false)
        }
      }
    },
    [loginAccount.email, loginAccount.password],
  )

  return (
    <Box
      width={{ base: '100%', md: '500px', '2xl': '700px' }}
      margin={{ base: '12px', md: '0px' }}
    >
      <LoginForm
        onSubmitForm={handleSubmitForm}
        onChangeForm={handleChange}
        emailError={errorMsgs.email}
        passwordError={errorMsgs.password}
        isDisable={isDisableButton}
      />
      {loading && <LoadingIndicator size="lg" />}
    </Box>
  )
}

export default LoginPage
