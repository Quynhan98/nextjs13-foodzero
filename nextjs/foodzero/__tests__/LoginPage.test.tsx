import React from 'react'
import Router from 'next/router'

// Utils
import {
  cleanup,
  fireEvent,
  render,
  RenderResult,
  screen,
  waitFor,
} from '@utils/testUtils'

// Constants
import {
  EMAIL_NOT_EXIST,
  INVALID_PASSWORD,
  PAGE_URL,
  SERVER_ERROR,
} from '@constants/index'

// Mocks
import { USERS_MOCK } from '@mocks/mockData'

// App
import LoginPage from 'app/(auth)/login/LoginPage'

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}))

describe('Login Page', () => {
  const accountMock = {
    email: 'foodzero@gmail.com',
    password: '123456789',
  }

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve(USERS_MOCK),
    }),
  ) as jest.Mock

  let wrapper: RenderResult

  beforeEach(() => {
    cleanup()

    wrapper = render(<LoginPage />)
  })

  it('Login Page matches snapshot', () => {
    const { container } = wrapper

    expect(container).toMatchSnapshot()
  })

  it('Should call onChange input email', () => {
    const email = screen.getByTestId('email')

    fireEvent.change(email, {
      target: { value: accountMock.email },
    })

    expect(email).toHaveValue(accountMock.email)
  })

  it('Should call onChange input password', () => {
    const password = screen.getByTestId('password')

    fireEvent.change(password, {
      target: { value: accountMock.password },
    })

    expect(password).toHaveValue(accountMock.password)
  })

  it('Should call submit form', () => {
    const email = screen.getByTestId('email')
    fireEvent.change(email, {
      target: { value: accountMock.email },
    })

    const password = screen.getByTestId('password')
    fireEvent.change(password, {
      target: { value: accountMock.password },
    })

    const button = screen.getByRole('button', {
      name: /Login/i,
    })

    fireEvent.submit(button)

    waitFor(() => {
      expect(Router.push).toHaveBeenCalledWith(PAGE_URL.HOME)
    })
  })

  it('Should call submit form error', () => {
    const email = screen.getByTestId('email')
    fireEvent.change(email, {
      target: { value: 'notemail@gmail.com' },
    })

    const password = screen.getByTestId('password')
    fireEvent.change(password, {
      target: { value: accountMock.password },
    })

    const button = screen.getByRole('button', {
      name: /Login/i,
    })
    fireEvent.submit(button)

    waitFor(() => {
      expect(screen.getByText(EMAIL_NOT_EXIST)).toBeTruthy()
      expect(screen.getByText(INVALID_PASSWORD)).toBeTruthy()
    })
  })

  it('Should call submit form with server error', () => {
    jest.fn().mockImplementationOnce(() => Promise.reject(SERVER_ERROR))

    const email = screen.getByTestId('email')
    fireEvent.change(email, {
      target: { value: accountMock.email },
    })

    const password = screen.getByTestId('password')
    fireEvent.change(password, {
      target: { value: accountMock.password },
    })

    const button = screen.getByRole('button', {
      name: /Login/i,
    })

    fireEvent.submit(button)

    waitFor(() => {
      expect(screen.getByText(SERVER_ERROR)).toBeTruthy()
    })
  })
})
