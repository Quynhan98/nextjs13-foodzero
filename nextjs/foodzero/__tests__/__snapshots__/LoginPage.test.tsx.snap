// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Login Page matches snapshot 1`] = `
<div>
  <div
    class="css-11vt9vz"
  >
    <form>
      <label
        class="chakra-form__label css-1vhhu1x"
      >
        Login
      </label>
      <div
        class="chakra-form-control css-xj5wq"
        role="group"
      >
        <input
          class="chakra-input css-1njo91c"
          data-testid="email"
          id="field-:r0:"
          name="email"
          placeholder="Email"
          type="email"
        />
        <input
          class="chakra-input css-1abompr"
          data-testid="password"
          id="field-:r0:"
          name="password"
          placeholder="Password"
          type="password"
        />
        <button
          class="chakra-button css-1copbg2"
          disabled=""
          type="submit"
        >
          Login
        </button>
      </div>
    </form>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
